lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -203,41 +210,42 @@ hasValidCodingKeysEnum(TypeChecker &tc, NominalTypeDecl *type,
 
   auto *codingKeysTypeDecl = dyn_cast<TypeDecl>(result);
   if (!codingKeysTypeDecl) {
-    // TODO: Produce a diagnostic complaining that the "CodingKeys" entity we
-    //       found is not a type.
+    tc.diagnose(result->getLoc(),
+                diag::codable_codingkeys_type_is_not_an_enum_here,
+                proto->getDeclaredType());
     return {/* has type? */ true, /* error? */ true};
   }
 
+  // CodingKeys may be a typealias. If so, follow the alias to its canonical
+  // type.
+  auto codingKeysType = codingKeysTypeDecl->getDeclaredInterfaceType();
+  if (isa<TypeAliasDecl>(codingKeysTypeDecl)) {
+    auto canType = codingKeysType->getCanonicalType();
+    assert(canType);
+    codingKeysTypeDecl = canType->getAnyNominal();
```

Comment:
Good to know. I'm preparing another PR with some additional compiler changes, so I can integrate this there.